src/utils/queue_builder.rs

```diff
@@ -1,146 +1,175 @@
 use crate::{
-    db::Pool, docbuilder::RustwideBuilder, utils::pubsubhubbub, BuildQueue, DocBuilder, Storage,
+    db::Pool, docbuilder::RustwideBuilder, utils::pubsubhubbub::HubPinger, BuildQueue, DocBuilder,
+    Storage,
 };
 use failure::Error;
 use log::{debug, error, info, warn};
-use std::panic::{catch_unwind, AssertUnwindSafe};
+use std::panic;
 use std::sync::Arc;
-use std::thread;
 use std::time::Duration;
+use tokio::{runtime::Handle, sync::Mutex, task, time};
 
+/// This is the core of the documentation building, within the inner loop we should never panic in any way
 // TODO: change to `fn() -> Result<!, Error>` when never _finally_ stabilizes
-pub fn queue_builder(
-    mut doc_builder: DocBuilder,
+pub async fn queue_builder(
+    doc_builder: DocBuilder,
     db: Pool,
     build_queue: Arc<BuildQueue>,
     storage: Arc<Storage>,
 ) -> Result<(), Error> {
-    /// Represents the current state of the builder thread.
-    enum BuilderState {
-        /// The builder thread has just started, and hasn't built any crates yet.
-        Fresh,
-        /// The builder has just seen an empty build queue.
-        EmptyQueue,
-        /// The builder has just seen the lock file.
-        Locked,
-        /// The builder has just finished building a crate. The enclosed count is the number of
-        /// crates built since the caches have been refreshed.
-        QueueInProgress(usize),
-    }
-
-    let mut builder = RustwideBuilder::init(db, storage)?;
-
-    let mut status = BuilderState::Fresh;
+    // These are all locked up due to the cross-thread sharing currently involved, it may be able to be removed after the
+    // build pipeline is async
+    let doc_builder = Arc::new(Mutex::new(doc_builder));
```

Comment:
I haven't checked these as closely, but looks like the same thing here, these are never locked concurrently so can just be a `sync::Mutex`?

Reply:
Locking is done across `.await` points, so using std or parking_lot's mutex makes the future `!Send`
```rust
error: future cannot be sent between threads safely
   --> src\utils\daemon.rs:90:5
    |
90  |     task::spawn(async move {
    |     ^^^^^^^^^^^ future created by async block is not `Send`
    |
   ::: C:\Users\Chase\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-0.2.22\src\task\spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::task::spawn::spawn`
    |
    = help: within `impl std::future::Future`, the trait `std::marker::Send` is not implemented for `*mut ()`
note: future is not `Send` as this value is used across an await
   --> src\utils\queue_builder.rs:51:29
    |
51  |             if let Err(e) = doc_builder.lock().load_cache().await {
    |                             ------------------^^^^^^^^^^^^^^^^^^^
    |                             |
    |                             await occurs here, with `doc_builder.lock()` maybe used later
    |                             has type `lock_api::mutex::MutexGuard<'_, parking_lot::raw_mutex::RawMutex, docbuilder::DocBuilder>` which is not `Send`
52  |                 error!("Failed to load cache: {}", e);
53  |             }
    |             - `doc_builder.lock()` is later dropped here
```